### DISCUSSION TITLE
Writing your own Micropython library

### DISCUSSION THREAD

**Discussion by roaldarbol** [BrianPugh/belay#108] [General]
Hi Brian,
I figured you would possibly be the right person to ask about this. I have this optic flow sensor, [ADNS-3080](https://www.amazon.co.uk/ADNS-3080-Optical-Horizontal-Movement-Detection/dp/B0728GGSSX/ref=sr_1_1?crid=2H0M1BVG0KZKR&keywords=iHaospace+ADNS-3080+Optical+Flow+Sensor+Module+APM2.52+APM2.6+Horizontal+Movement+Detection&qid=1676453096&s=electronics&sprefix=%2Celectronics%2C1002&sr=1-1), that I can't find a Micropython library for. I know there are Arduino libraries out there, and that it uses SPI, but that's about it. I have no idea how to go about writing a library myself, but I figured you might know how to. ;-) (but now worries if not). I was thinking there might be a way using the [`SPI` class from `machine`](https://docs.micropython.org/en/latest/library/machine.SPI.html), but that's where my brain stops. (There's talk of `0x42` which I don't really know about).
So I was wondering if you have some suggestions on where to start, if it's worth the time, or some suggested materials. 
Thanks!

**Comment by BrianPugh** [comment]
So first step: ask ChatGPT 😄. It responds with:

```
import time
from machine import Pin, SPI

class ADNS3080:
    def __init__(self, spi, cs_pin):
        self.spi = spi
        self.cs = Pin(cs_pin, Pin.OUT)
        self.cs.on()
        self.reset()
        
    def reset(self):
        self.cs.off()
        self.spi.write(b"\x3A")
        self.cs.on()
        time.sleep_ms(50)
        
    def read_register(self, addr):
        self.cs.off()
        self.spi.write(bytes([addr | 0x80]))
        value = self.spi.read(1)
        self.cs.on()
        return value[0]
    
    def write_register(self, addr, value):
        self.cs.off()
        self.spi.write(bytes([addr]))
        self.spi.write(bytes([value]))
        self.cs.on()
    
    def read_motion(self):
        self.write_register(0x02, 0x01) # take image
        while not (self.read_register(0x03) & 0x80): # wait for motion
            time.sleep_us(100)
        dx = self.read_register(0x04)
        dy = self.read_register(0x05)
        self.write_register(0x02, 0x00) # stop motion
        return (dx, dy)

# set up SPI interface
spi = SPI(1, baudrate=5000000, polarity=0, phase=0)
cs_pin = 0 # change to the appropriate pin for your setup
adns = ADNS3080(spi, cs_pin)

while True:
    dx, dy = adns.read_motion()
    print("dx = ", dx, "dy = ", dy)
    time.sleep(0.1)
```

Some corrections that I think need to be made to ChatGPT's (above) code:

* According to the datasheet, I think all of the specified registers are off by one:
  * Motion register should be `0x02`, not `0x03`
  * According to the datasheet, the address for Delta_X is `0x03`, not `0x04`.
  * According to the datasheet, the address for Delta_Y is `0x04`, not `0x05`.
  
 It's probably wrong in several other ways, but its a pretty good starting point. I would start from the above code, and sort of [compare it against an existing repo](https://github.com/Lauszus/ADNS3080/blob/master/ADNS3080.ino) as well as the datasheet.
 
 Side note: For my classes that take in an `spi` object, like [the Max6675 driver I wrote](https://github.com/BrianPugh/micropython-libs/blob/main/lib/max6675.py), I like to set the baudrate before each command, then restore it after. This is so that a bunch of different devices can share the same SPI bus, even if they don't all operate at the same speed. The baudrate restore code is a bit jank since it's not explicitly exposed by micropython; it's also unnecessary if all device classes set their own desired baudrate.

**Comment by BrianPugh** [comment]
As for testing the code, probably the quickest way would be to run `belay run <PORT> my_driver.py`

**Comment by roaldarbol** [comment]
Dude, that's amazing - thanks a ton! I'll give it a shot and report back. ChatGPT is a f*ing gamechanger. :P
